## NUMPY 
##### [Visit NumPy Documentation](https://numpy.org/doc/)
---

In [1]:
import numpy as np
import sys

## Hands on !!

In [2]:
# basic list in python
arr = [1,2,3,4]
arr1 = [0,0.5,1,1.5,2]

In [3]:
#  converting it into a numpy array
a = np.array(arr)
b = np.array(arr1)

In [4]:

type(a)
type(b)

numpy.ndarray

## numpy operation 
---

### numpy array indexing and slicing :

In [5]:

a[0], a[1], a[-1]

(1, 2, 4)

In [6]:

a[1:2]

array([2])

In [7]:
a[:2]

array([1, 2])

In [8]:
a[-3:]

array([2, 3, 4])

In [9]:
# same like python list
# a[start,end,steps]
a[0:-1:2]

array([1, 3])

In [10]:
a[::2]

array([1, 3])

In [11]:
# second array
b[0], b[2], b[-1]

(0.0, 1.0, 2.0)

In [12]:
# it means that from b you wanna select elements from index 0, 2, -1 resp.
b[[0 , 2, -1]]

array([0., 1., 2.])

## Array types
---

In [13]:
a

array([1, 2, 3, 4])

In [14]:
# note that dtype is used to check the data type of the numpy array
# note that dtype is not same as the type() in general python rather it is specially for numpy
a.dtype

dtype('int32')

In [15]:
b

array([0. , 0.5, 1. , 1.5, 2. ])

In [16]:
# note that b has floating val in the array hence datatype is float
b.dtype

dtype('float64')

In [17]:
# we can change the data type of the numpy array
# note that np.float is depreciated hence use only float or np.float64
np.array([1,2,3,4], dtype=float)

array([1., 2., 3., 4.])

In [18]:
# same case for int 
np.array([1.0,2.1,3.2,5.4],dtype=int)

array([1, 2, 3, 5])

In [19]:
# note that numpy store int , float , boolean but not string in a regular way how ever they can be stored in a unique format
# note that numpy stores string / characters in unicode or byte strings hence dtype=<Ux
# where x represent the longest string present in the numpy array
np.array(['a','b','cat','d'])

array(['a', 'b', 'cat', 'd'], dtype='<U3')

In [20]:
# NumPy prefers homogeneous arrays, but here you're mixing a dictionary ({'a': 1}) and (presumably) a module (sys). 
# NumPy will default to dtype=object in such cases.
d = np.array([{'a': 1}, sys])  # Creates an object dtype array
print(d)
print(d.dtype)  # Output: object

[{'a': 1} <module 'sys' (built-in)>]
object


## Dimension and Shapes
---

### 2D array

In [21]:
A = np.array([
    [1,2,3],
    [5,6,7]
])
A

array([[1, 2, 3],
       [5, 6, 7]])

In [22]:
# shape of the arrays / it means length breadth height 
A.shape

(2, 3)

In [23]:
# dimensions of the array : x , y , z axis / # dimension of the array : 2d or 3d
A.ndim

2

In [24]:
# total size of the array : total no. of elements
A.size

6

### 3D array

In [25]:
B = np.array([
    [
        [1,2,3],
        [4,5,6]
    ],
    [
        [7,8,9],
        [10,11,12]
    ]
])

In [26]:
B

array([[[ 1,  2,  3],
        [ 4,  5,  6]],

       [[ 7,  8,  9],
        [10, 11, 12]]])

In [27]:
# shape of the array : it means length breadth height 
B.shape

(2, 2, 3)

In [28]:
# dimension of the array : 2d or 3d
B.ndim

3

In [29]:
B.size

12

In [30]:
# note that if the numpy array either 2d or 3d is not in the proper alignment then the python consider it as the normal python object

### Indexing and Slicing of matrix

In [31]:
# square matrix
A = np.array(
    [
        [1,2,3],
        [4,5,6],
        [7,8,9]
    ]
)

In [32]:
# accessing 1 dimension of the array
A[0]

array([1, 2, 3])

In [33]:
A[1]

array([4, 5, 6])

In [34]:
A[2]

array([7, 8, 9])

In [35]:
A[0:3]

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [36]:
# accessing 2 dimension of the array
A[0][2]

3

In [37]:
A[1][2]

6

In [38]:
# splicing in both 2d and 3d
# array_name[dim1_slicing, dim2_slicing]
A[0:3, 1:3]

array([[2, 3],
       [5, 6],
       [8, 9]])

In [39]:
# changing the value of the index
A[1] = np.array([10,10,10])

In [40]:
A

array([[ 1,  2,  3],
       [10, 10, 10],
       [ 7,  8,  9]])

In [41]:
# if we assign a single no. to a incompatible array shape then that no. id repeated times that dimension
A[2] = 99

In [42]:
A

array([[ 1,  2,  3],
       [10, 10, 10],
       [99, 99, 99]])

### Built-In function in Numpy

#### for 1D array

In [43]:
a = np.array([1,2,3,4])

In [44]:
# function to find the sum of the elements of the array
a.sum()

10

In [45]:
# function to find the mean of the elements of the array
a.mean()

2.5

#### standard deviation 
---
##### The standard deviation (σ) is calculated as:
## σ = √(∑(xi−mu)^2/n)

In [46]:
# function to find the standard deviation of the numpy 
a.std()

1.118033988749895

#### variance
---
##### variance is calculated as :

## σ^2 =  1/N (∑(i = 1 to N) of (xi - mu)^2)

In [47]:
# function to find the variance of the array
a.var()

1.25

#### for multi-d array

In [48]:
A = np.array(
    [
        [1,2,3],
        [4,5,6],
        [7,8,9]
    ]
)

In [49]:
A.sum()

45

In [50]:
A.mean()

5.0

In [51]:
A.std()

2.581988897471611

In [52]:
A.var()

6.666666666666667

#### above function but along an axis -> x or y

note that here : 
- axis = 0 means operation along column
- axis = 1 means operation along row

In [53]:
# sum along column
A.sum(axis = 0) 

array([12, 15, 18])

In [54]:
# sum along column
A.sum(axis = 1)

array([ 6, 15, 24])

In [55]:
# mean along column
A.mean(axis = 0)

array([4., 5., 6.])

In [56]:
# mean along row
A.mean(axis = 1)

array([2., 5., 8.])

In [57]:
# std along col
A.std(axis = 0)

array([2.44948974, 2.44948974, 2.44948974])

In [58]:
# std along row
A.std(axis = 1)

array([0.81649658, 0.81649658, 0.81649658])

In [59]:
# var along col
A.var(axis = 0)

array([6., 6., 6.])

In [60]:
# var along row
A.var(axis = 1)

array([0.66666667, 0.66666667, 0.66666667])

### Broadcasting and vectorized Operation
- vectorized operations are operations which are performed between arr & arr || arr & scalar values

#### Array (operation) scalar val

In [61]:
# .arange(n) functions created an array of numbers from 0 to n-1
a = np.arange(4)
a

array([0, 1, 2, 3])

In [62]:
# adding 10 to array a
# arr + 10 = add 10 to each elements of the array
a + 10

array([10, 11, 12, 13])

In [63]:
# multiply 10 tp array a
# arr * 10 = multiply 10 to each elements of the array
a * 10


array([ 0, 10, 20, 30])

In [64]:
# a is unchanged
a

array([0, 1, 2, 3])

In [65]:
# this changes the array itself
a += 100

In [66]:
a

array([100, 101, 102, 103])

In [67]:
# loading a list using for loop
l = [i for i in range(4)]
l

[0, 1, 2, 3]

In [68]:
l = [ i * 10 for i in l]
l

[0, 10, 20, 30]

#### array (operation) array
- note that the arrays needs to have compatible shape for vectorised operations
- NumPy Operations: Element-wise vs. Matrix Operations

| **Operation**             | **Symbol**                      | **Default Behavior**                                      |
|---------------------------|--------------------------------|----------------------------------------------------------|
| **Addition**              | `+`                            | Element-wise                                            |
| **Subtraction**           | `-`                            | Element-wise                                            |
| **Multiplication**        | `*`                            | Element-wise                                            |
| **Division**              | `/`                            | Element-wise                                            |
| **Exponentiation**        | `**`                           | Element-wise                                            |

In [69]:
# creating a new array
a = np.arange(4)
a

array([0, 1, 2, 3])

In [70]:
b = np.array([10,10,10,10])
b

array([10, 10, 10, 10])

In [71]:
# element wise addition
a + b

array([10, 11, 12, 13])

In [72]:
# element wise multiplication
a * b

array([ 0, 10, 20, 30])

#### Boolean array
_also called mask_

##### 1-D Array

In [73]:
a = np.arange(4)
a

array([0, 1, 2, 3])

In [74]:
# selecting element using index
a[[0, -1]]

array([0, 3])

In [75]:
# selecting elements using boolean value
# note that true means to take the element and false means to not take the elements
a[[True, False, False, True]]

array([0, 3])

In [76]:
a

array([0, 1, 2, 3])

In [77]:
# here select elements >= 2 returning a boolean array
a >= 2

array([False, False,  True,  True])

In [78]:
# filters the array on a[condition] returning the array of elements
# note that in [] a is the element of the array a
a[a>=2]

array([2, 3])

In [79]:
a.mean()

1.5

In [80]:
a[a>a.mean()]

array([2, 3])

In [81]:
a[~(a > a.mean())]

array([0, 1])

In [82]:
a[(a == 0 )|(a == 1)]

array([0, 1])

In [83]:
a[(a <= 2) & (a % 2 == 0)]

array([0, 2])

##### Multi-D array

In [84]:
A = np.random.randint(100,size = (3,3))
A

array([[27, 82, 94],
       [58, 64, 12],
       [15, 84, 17]])

In [85]:
# now creating a boolean array based on the above array returning a boolean 1D array
A[np.array([
    [True, False, True],
    [False, True, False],
    [True, False, True]
])]

array([27, 94, 64, 15, 17])

In [86]:
# this is dynamically generated boolean array while filtering
A > 30

array([[False,  True,  True],
       [ True,  True, False],
       [False,  True, False]])

---
### Linear Algebra methods

| **Operation**                     | **Function**                        | **Description** |
|------------------------------------|-------------------------------------|----------------|
| **Matrix Multiplication**         | `np.dot(A, B)` or `A @ B`          | Performs **matrix multiplication** following linear algebra rules. |
| **Element-wise Multiplication**   | `A * B`                             | Multiplies matrices **element-wise** (Hadamard product). |
| **Matrix Power**                   | `np.linalg.matrix_power(A, n)`      | Raises a square matrix `A` to the power `n`. |
| **Determinant**                    | `np.linalg.det(A)`                  | Computes the **determinant** of matrix `A`. |
| **Inverse**                        | `np.linalg.inv(A)`                  | Computes the **inverse** of matrix `A`. |
| **Pseudo-Inverse**                 | `np.linalg.pinv(A)`                 | Computes the **Moore-Penrose pseudoinverse** of `A`. |
| **Transpose**                      | `A.T` or `np.transpose(A)`          | Returns the **transpose** of matrix `A`. |
| **Trace**                          | `np.trace(A)`                       | Returns the **sum of diagonal elements** of matrix `A`. |
| **Eigenvalues & Eigenvectors**     | `np.linalg.eig(A)`                  | Computes **eigenvalues** and **eigenvectors** of `A`. |
| **Singular Value Decomposition**   | `np.linalg.svd(A)`                  | Computes the **Singular Value Decomposition (SVD)**. |
| **QR Decomposition**               | `np.linalg.qr(A)`                   | Computes the **QR decomposition** of `A`. |
| **Cholesky Decomposition**         | `np.linalg.cholesky(A)`             | Computes the **Cholesky decomposition** of `A`. |
| **LU Decomposition**               | `scipy.linalg.lu(A)`                | Computes the **LU decomposition** (requires SciPy). |
| **Solve Linear System**            | `np.linalg.solve(A, b)`             | Solves the equation `Ax = b` for `x`. |
| **Least Squares Solution**         | `np.linalg.lstsq(A, b, rcond=None)` | Solves the **least squares problem** for `Ax = b`. |
| **Norm of a Matrix**               | `np.linalg.norm(A, ord)`            | Computes the **norm** of a matrix or vector (Frobenius, Euclidean, etc.). |
| **Condition Number**               | `np.linalg.cond(A)`                 | Computes the **condition number** of `A`. |
| **Rank of a Matrix**               | `np.linalg.matrix_rank(A)`          | Computes the **rank** of matrix `A`. |
| **Outer Product**                  | `np.outer(a, b)`                    | Computes the **outer product** of vectors `a` and `b`. |
| **Inner Product**                  | `np.inner(a, b)`                    | Computes the **inner product** of vectors `a` and `b`. |
| **Cross Product**                  | `np.cross(a, b)`                    | Computes the **cross product** of 3D vectors `a` and `b`. |



In [87]:
A = np.array([
    [1,2,3],
    [4,5,6],
    [7,8,9]
])

In [88]:
B = np.array([
    [6,5],
    [4,3],
    [2,1]
])

In [89]:
# here it is called matrix multiplication
A.dot(B)

array([[20, 14],
       [56, 41],
       [92, 68]])

In [90]:
# here it is also called matrix multiplication
A @ B

array([[20, 14],
       [56, 41],
       [92, 68]])

In [91]:
# transpose of a matrix
B.T

array([[6, 4, 2],
       [5, 3, 1]])

In [92]:
A.T

array([[1, 4, 7],
       [2, 5, 8],
       [3, 6, 9]])

In [93]:
# B transpose X A
B.T @ A

array([[36, 48, 60],
       [24, 33, 42]])

---
### Size of object in memory
##### int, float
- the total no. of bytes needed to store a simple number
- Conclusion : Not at all space efficient

In [94]:
# An integer in python is > 24 bytes
sys.getsizeof(1)

28

In [95]:
# Longs are even larger
sys.getsizeof(10**100)

72

In [96]:
# Numpy size is much smaller
np.dtype(int).itemsize

4

In [97]:
# numpy size is much smaller
np.dtype(np.int8).itemsize

1

In [98]:
np.dtype(float).itemsize

8

#### List size
##### list are even larger
- here we are comparing the size of python list and numpy arrays

In [99]:
# size of 1-element list in python
sys.getsizeof([1])

64

In [100]:
# size of 1-element array in numpy
np.array([1]).nbytes

4

#### Performance is also important
- here we are comparing the performance of the python list and numpy arrays

In [101]:
l = list(range(1000))

In [102]:
a = np.arange(1000)

##### here we will be using the time module to compare the performance

In [108]:
# here we are calculating the time to calculate the exponent of each element in a
%timeit np.sum(a ** 2)

3.83 µs ± 134 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [109]:
# here we are calculating the time to calculate the exponent of each element in b
%timeit sum([i**2 for i in l])

193 µs ± 2.81 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


##### conclusion on the above experiment : 
#### 🔍 **What Do These Results Mean?**

- **For NumPy array:** `3.83 µs ± 134 ns per loop` (very fast 🔥)  
- **For Python list:** `193 µs ± 2.81 µs per loop` (much slower 🐢)  

✅ **NumPy is ~50 times faster than a Python list for this operation!**


---
### Numpy more useful functions 

#### random

In [119]:
# generates two random numbers between 0 and 1 from a uniform distribution.
np.random.random(size=2)

array([0.43752093, 0.00453949])

In [120]:
# Generates values from a normal distribution of 2 elements
np.random.normal(size=2)

array([ 2.06736737, -1.62655039])

In [121]:
# Generates a random m x n array of numbers between 0 and 1
# random.rand(m,n)
np.random.rand(2,4)

array([[0.12433965, 0.15977823, 0.67709256, 0.50344444],
       [0.76963128, 0.53831063, 0.69718689, 0.60894567]])

#### arange

In [122]:
# generate a numpy array containing element from 0 to 9
np.arange(10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [124]:
# generate a numpy array containing element from m to n
np.arange(5,10)

array([5, 6, 7, 8, 9])

In [125]:
# Creates an array with values from 0 to 1 with a step of .1
np.arange(0,1,.1)

array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])

#### reshape

In [128]:
# Reshapes an array into a specified shape ( row,col)
np.arange(10).reshape(2,5)

array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9]])

In [129]:
# Reshapes an array into a specified shape (row,col)
np.arange(10).reshape(5,2)

array([[0, 1],
       [2, 3],
       [4, 5],
       [6, 7],
       [8, 9]])

#### linspace

In [131]:
#  generate an array of evenly spaced numbers between a start and stop value and no. of val to generate
np.linspace(0,1,5)

array([0.  , 0.25, 0.5 , 0.75, 1.  ])

In [139]:
np.linspace(1,100,20,dtype = int)

array([  1,   6,  11,  16,  21,  27,  32,  37,  42,  47,  53,  58,  63,
        68,  73,  79,  84,  89,  94, 100])

In [140]:
# endpoint=False → Excludes the stop value (1), so spacing is adjusted accordingly
np.linspace(0,1,20,False)

array([0.  , 0.05, 0.1 , 0.15, 0.2 , 0.25, 0.3 , 0.35, 0.4 , 0.45, 0.5 ,
       0.55, 0.6 , 0.65, 0.7 , 0.75, 0.8 , 0.85, 0.9 , 0.95])

#### Zeros, ones, empty

In [142]:
# Return a new array of given shape and type, filled with zeros.
np.zeros(5)

array([0., 0., 0., 0., 0.])

In [144]:
np.zeros((3,3))

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

In [145]:
np.zeros((3,3),dtype=int)

array([[0, 0, 0],
       [0, 0, 0],
       [0, 0, 0]])

In [147]:
# Return a new array of given shape and type, filled with ones.
np.ones(5)

array([1., 1., 1., 1., 1.])

In [148]:
np.ones((3,3))

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

In [153]:
# Return a new array of given shape and type, without initializing entries.
# It does not initialize values, meaning the elements will contain random garbage values (whatever was in memory at that time).
# It's faster than np.zeros() or np.ones() because it skips the step of setting values.
np.empty(5)

array([1., 1., 1., 1., 1.])

In [154]:
np.empty((2,5))

array([[9.43940533e-312, 1.77863633e-322, 0.00000000e+000,
        0.00000000e+000, 0.00000000e+000],
       [1.09322382e-042, 7.48418037e-067, 4.76066267e-090,
        2.21657816e-056, 4.34339413e-315]])

#### identity, eye

In [163]:
# Return the identity array.
# alaways generetes square matrix
np.identity(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [167]:
# Return a 2-D array with ones on the diagonal and zeros elsewhere.
# rows may not be same as col
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [168]:
np.eye(3,3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [169]:
np.eye(8,4)

array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [171]:
# The k parameter controls which diagonal gets the 1s. k is called the offset
np.eye(8,4,k=1)

array([[0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [176]:
# +ve k -> shift col , -ve k -> shift row
np.eye(8,4,k=-3)

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.],
       [0., 0., 0., 0.]])

In [177]:
"hello World"[6]

'W'

# 📌 NumPy Functions Table

## **Creating Arrays**
| Function | Syntax | Use Case |
|----------|--------|----------|
| `array()` | `np.array([1, 2, 3])` | Creates a NumPy array from a list or tuple |
| `zeros()` | `np.zeros((3,3))` | Creates an array filled with zeros |
| `ones()` | `np.ones((2,2))` | Creates an array filled with ones |
| `full()` | `np.full((3,3), 5)` | Creates an array filled with a specific value |
| `eye()` | `np.eye(3)` | Creates an identity matrix of size `3×3` |
| `arange()` | `np.arange(1, 10, 2)` | Creates an array with values from 1 to 10 with a step of 2 |
| `linspace()` | `np.linspace(0, 1, 5)` | Creates an array with 5 equally spaced values between 0 and 1 |
| `random.rand()` | `np.random.rand(3,3)` | Generates a random `3×3` array of numbers between 0 and 1 |

## **Mathematical Operations**
| Function | Syntax | Use Case |
|----------|--------|----------|
| `add()` | `np.add(a, b)` | Element-wise addition of two arrays |
| `subtract()` | `np.subtract(a, b)` | Element-wise subtraction |
| `multiply()` | `np.multiply(a, b)` | Element-wise multiplication |
| `divide()` | `np.divide(a, b)` | Element-wise division |
| `power()` | `np.power(a, 2)` | Raises each element of `a` to the power of 2 |
| `sqrt()` | `np.sqrt(a)` | Computes the square root of each element in `a` |
| `exp()` | `np.exp(a)` | Computes the exponential (e^x) for each element in `a` |
| `log()` | `np.log(a)` | Computes the natural logarithm (ln) of `a` |

## **Statistical Functions**
| Function | Syntax | Use Case |
|----------|--------|----------|
| `mean()` | `np.mean(a)` | Computes the mean of an array |
| `median()` | `np.median(a)` | Computes the median of an array |
| `std()` | `np.std(a)` | Computes the standard deviation |
| `var()` | `np.var(a)` | Computes the variance |
| `min()` | `np.min(a)` | Finds the minimum value in an array |
| `max()` | `np.max(a)` | Finds the maximum value in an array |
| `argmin()` | `np.argmin(a)` | Returns the index of the minimum value |
| `argmax()` | `np.argmax(a)` | Returns the index of the maximum value |

## **Linear Algebra**
| Function | Syntax | Use Case |
|----------|--------|----------|
| `dot()` | `np.dot(a, b)` | Computes the dot product of two arrays |
| `matmul()` | `np.matmul(a, b)` | Matrix multiplication |
| `linalg.inv()` | `np.linalg.inv(A)` | Computes the inverse of a matrix |
| `linalg.det()` | `np.linalg.det(A)` | Computes the determinant of a matrix |
| `linalg.eig()` | `np.linalg.eig(A)` | Computes eigenvalues and eigenvectors |

## **Array Manipulation**
| Function | Syntax | Use Case |
|----------|--------|----------|
| `reshape()` | `np.reshape(a, (2,3))` | Reshapes an array into a specified shape |
| `ravel()` | `np.ravel(a)` | Flattens an array into a 1D array |
| `transpose()` | `np.transpose(a)` | Transposes a matrix |
| `concatenate()` | `np.concatenate((a, b), axis=0)` | Joins two arrays along a specified axis |
| `vstack()` | `np.vstack((a, b))` | Stacks arrays vertically |
| `hstack()` | `np.hstack((a, b))` | Stacks arrays horizontally |

## **Sorting & Searching**
| Function | Syntax | Use Case |
|----------|--------|----------|
| `sort()` | `np.sort(a)` | Sorts an array |
| `argsort()` | `np.argsort(a)` | Returns the indices that would sort an array |
| `where()` | `np.where(a > 5, 1, 0)` | Returns indices where a condition is met |
| `unique()` | `np.unique(a)` | Finds unique values in an array |

## **Random Number Generation**
| Function | Syntax | Use Case |
|----------|--------|----------|
| `random.rand()` | `np.random.rand(3,3)` | Generates a random `3×3` array (values between 0 and 1) |
| `random.randint()` | `np.random.randint(1, 10, (2,2))` | Generates a random integer matrix (between 1 and 10) |
| `random.normal()` | `np.random.normal(0, 1, (3,3))` | Generates values from a normal distribution |
| `random.choice()` | `np.random.choice(a, 3)` | Selects 3 random elements from an array |
| `random.shuffle()` | `np.random.shuffle(a)` | Shuffles an array in place |

---

## **🔹 Summary**
- **Creating arrays:** `array()`, `zeros()`, `ones()`, `arange()`, `linspace()`
- **Math operations:** `add()`, `subtract()`, `multiply()`, `divide()`, `sqrt()`
- **Statistics:** `mean()`, `median()`, `std()`, `var()`, `min()`, `max()`
- **Linear algebra:** `dot()`, `matmul()`, `inv()`, `det()`, `eig()`
- **Array manipulation:** `reshape()`, `transpose()`, `concatenate()`
- **Sorting & Searching:** `sort()`, `argsort()`, `where()`, `unique()`
- **Random generation:** `rand()`, `randint()`, `choice()`, `shuffle()`

🚀 **NumPy is incredibly powerful!** Let me know if you need more details! 😃


-----------------------------------------------------------**END**---------------------------------------------------